File: proxy/src/client/task.ts
import { credentials } from '@grpc/grpc-js'
import { TasksClient } from '../protobuf/tasks_grpc_pb'
import {
  CreateTaskRequest,
  CreateTaskReply,
  GetTaskListRequest,
  GetTaskListReply,
  GetTaskReply,
  GetTaskRequest,
  UpdateTaskRequest,
  UpdateTaskReply
} from '../protobuf/tasks_pb'

const createClient = (): TasksClient => {
  return new TasksClient(
    'localhost:8080',
    credentials.createInsecure(),
  );
}

export const create = (params: CreateTaskRequest.AsObject): Promise<CreateTaskReply.AsObject> => {
  const request = new CreateTaskRequest();
  const client = createClient()

  request.setUid(params.uid);
  request.setTitle(params.title);
  request.setContent(params.content);

  return new Promise((resolve, reject) => {
    client.createTask(request, (error, response) => {
      if (error) {
        console.error(error);
        reject({
          code: error.code || 500,
          message: error.message || "something went wrong",
        });
      }

      return resolve(response.toObject());
    });
  });
}

export const getList = (params: GetTaskListRequest.AsObject): Promise<GetTaskListReply.AsObject> => {
  const request = new GetTaskListRequest()
  const client = createClient()

  request.setUid(params.uid)

  return new Promise((resolve, reject) => {
    client.getTaskList(request, (error, response) => {
      if (error) {
        // console.error(error);
        return reject({
          code: error.code || 500,
          message: error.message || "something went wrong",
        });
      }

      return resolve(response.toObject());
    });
  });
}

export const getById = (params: GetTaskRequest.AsObject): Promise<GetTaskReply.AsObject> => {
  const request = new GetTaskRequest()
  const client = createClient()

  request.setUid(params.uid)
  request.setId(params.id)

  return new Promise((resolve, reject) => {
    client.getTask(request, (error, response) => {
      if (error) {
        console.error(error);
        reject({
          code: error.code || 500,
          message: error.message || "something went wrong",
        });
      }

      return resolve(response.toObject());
    });
  });
}

export const update = (params: UpdateTaskRequest.AsObject): Promise<UpdateTaskReply.AsObject> => {
  const request = new UpdateTaskRequest()
  const client = createClient()

  request.setUid(params.uid)
  request.setId(params.id)
  request.setTitle(params.title)
  request.setContent(params.content)

  return new Promise((resolve, reject) => {
    client.updateTask(request, (error, response) => {
      if (error) {
        return reject({
          code: error.code || 500,
          message: error.message || "something went wrong",
        });
      }

      return resolve(response.toObject());
    });
  });
}

File: app/handler/task.go
package handler

import (
	"context"
	"time"

	"github.com/highhi/sandbox-grpc-go-node/app/domain/task"
	"github.com/highhi/sandbox-grpc-go-node/app/infra/persistence"
	pb "github.com/highhi/sandbox-grpc-go-node/app/protobuf"
)

type taskHandler struct {
	r *persistence.TaskRepository
	pb.UnimplementedTasksServer
}

func newTaskHandler(r *persistence.TaskRepository) *taskHandler {
	return &taskHandler{r: r}
}

func (h *taskHandler) CreateTask(ctx context.Context, in *pb.CreateTaskRequest) (*pb.CreateTaskReply, error) {
	now := time.Now()

	err := h.r.Create(&task.Task{
		UID:       in.GetUid(),
		Title:     in.GetTitle(),
		Content:   in.GetContent(),
		CreatedAt: now,
		UpdatedAt: now,
	})

	if err != nil {
		return nil, err
	}

	return &pb.CreateTaskReply{}, nil
}

func (h *taskHandler) GetTaskList(ctx context.Context, in *pb.GetTaskListRequest) (*pb.GetTaskListReply, error) {
	tasks, err := h.r.List(in.GetUid())
	if err != nil {
		return nil, err
	}
	return &pb.GetTaskListReply{Tasks: task.ToPBList(tasks)}, nil
}

func (h *taskHandler) GetTask(ctx context.Context, in *pb.GetTaskRequest) (*pb.GetTaskReply, error) {
	task, err := h.r.GetById(in.GetUID(), in.GetID())
	if err != nil {
		return nil, err
	}
	return &pb.GetTaskReply{Task: task.ToPB()}, nil
}

func (h *taskHandler) UpdateTask(ctx context.Context, in *pb.UpdateTaskRequest) (*pb.UpdateTaskReply, error) {
	task, err := h.r.GetById(in.GetUID(), in.GetID())

	if err != nil {
		return nil, err
	}

	task.Title = in.GetTitle()
	task.Content = in.GetContent()

	err = h.r.Update(task)
	if err != nil {
		return nil, err
	}
	return &pb.UpdateTaskReply{}, nil
}

func (h *taskHandler) DeleteTask(ctx context.Context, in *pb.DeleteTaskRequest) (*pb.DeleteTaskReply, error) {
	task, err := h.r.GetById(in.GetUID(), in.GetID())

	if err != nil {
		return nil, err
	}

	err = h.r.Delete(task)
	if err != nil {
		return nil, err
	}
	return &pb.DeleteTaskReply{}, nil
}
